src/cryptonote_core/quorum_cop.cpp

```diff
@@ -0,0 +1,210 @@
+    : m_core(core), m_service_node_list(service_node_list), m_last_height(0)
+  {
+  }
+
+  void quorum_cop::blockchain_detached(uint64_t height)
+  {
+    uint64_t delta_height = std::abs(static_cast<int64_t>(m_last_height - height));
+    if (delta_height > REORG_SAFETY_BUFFER_IN_BLOCKS)
+    {
+      LOG_ERROR("The blockchain was detached, quorum cop has processed votes for: " << delta_height <<
+                " blocks which is greater than the recommended REORG_SAFETY_BUFFER_IN_BLOCKS: " << REORG_SAFETY_BUFFER_IN_BLOCKS);
+    }
+
+    m_last_height = (height < REORG_SAFETY_BUFFER_IN_BLOCKS) ? height : height - REORG_SAFETY_BUFFER_IN_BLOCKS;
+  }
+
+  static bool participates_in_quorum(const std::vector<crypto::public_key>& quorum, crypto::public_key const &my_key, size_t *index_in_quorum = nullptr)
+  {
+    size_t index = 0;
+    for (crypto::public_key const &quorum_key : quorum)
+    {
+      if (quorum_key == my_key) break;
+      ++index;
+    }
+
+    if (index_in_quorum)
+      *index_in_quorum = index;
+
+    bool result = (index < quorum.size());
+    return result;
+  }
+
+  void quorum_cop::block_added(const cryptonote::block& block, const std::vector<cryptonote::transaction>& txs)
+  {
+    crypto::public_key my_pubkey;
+    crypto::secret_key my_seckey;
+    if (!m_core.get_service_node_keys(my_pubkey, my_seckey))
+      return;
+
+    time_t const now         = time(nullptr);
+    bool been_alive_for_2hrs = (now - m_core.get_start_time()) >= (60 * 60 * 2);
+    if (!been_alive_for_2hrs)
+    {
+      return;
+    }
+
+    uint64_t const height        = cryptonote::get_block_height(block);
+    uint64_t const latest_height = m_core.get_current_blockchain_height();
+
+    if (latest_height < loki::service_node_deregister::VOTE_LIFETIME_BY_HEIGHT)
+      return;
+
+    uint64_t const execute_justice_from_height = latest_height - loki::service_node_deregister::VOTE_LIFETIME_BY_HEIGHT;
+    if (height < execute_justice_from_height)
+      return;
+
+    if (m_last_height < execute_justice_from_height)
+      m_last_height = execute_justice_from_height;
+
+
+    for (;m_last_height < (height - REORG_SAFETY_BUFFER_IN_BLOCKS); m_last_height++)
+    {
+      const std::shared_ptr<quorum_state> state = m_core.get_quorum_state(m_last_height);
+      if (!state)
+      {
+        // TODO(loki): Fatal error
+        LOG_ERROR("Quorum state for height: " << m_last_height << "was not cached in daemon!");
+        continue;
+      }
+
+      size_t my_index_in_quorum = 0;
+      if (!participates_in_quorum(state->quorum_nodes, my_pubkey, &my_index_in_quorum))
+      {
+        continue;
+      }
+
+      for (size_t node_index = 0; node_index < state->nodes_to_test.size(); ++node_index)
+      {
+        const crypto::public_key &node_key = state->nodes_to_test[node_index];
+
+        CRITICAL_REGION_LOCAL(m_lock);
+        bool vote_off_node = (m_uptime_proof_seen.find(node_key) != m_uptime_proof_seen.end());
```

Comment:
Changed to just find the key, since the entries in the list should be under UPTIME_PROOF_MAX_IN_SECONDS +- 30seconds (depending on the pruner being called every 30s on idle).

Reply:
I would add a buffer. Right now there is a chance that we will drop the ping from 1 hour ago, and then this evaluation happens before the next ping arrives.

Reply:
Yeah sounds good